components/psst/browser/content/psst_tab_web_contents_observer.cc

```diff
@@ -199,8 +206,31 @@ void PsstTabWebContentsObserver::OnPolicyScriptResult(
     return;
   }
 
-  ui_delegate_->UpdateTasks(script_result_parsed->progress,
-                            script_result_parsed->applied_tasks);
+  timeout_timer_.Stop();
+  ui_delegate_->UpdateTasks(
+      script_result_parsed->progress, script_result_parsed->applied_tasks,
+      script_result_parsed->progress == 100 ? mojom::PsstStatus::kCompleted
+                                            : mojom::PsstStatus::kInProgress);
+}
+
+void PsstTabWebContentsObserver::RunWithTimeout(
+    const int last_committed_entry_id,
+    const std::string& script,
+    InsertScriptInPageCallback callback) {
+  timeout_timer_.Start(
+      FROM_HERE, kScriptTimeout,
+      base::BindOnce(&PsstTabWebContentsObserver::OnScriptTimeout,
+                     weak_factory_.GetWeakPtr(), last_committed_entry_id));
+  inject_script_callback_.Run(script, std::move(callback));
+}
+
+void PsstTabWebContentsObserver::OnScriptTimeout(int id) {
+  if (!ShouldInsertScriptForPage(id)) {
+    return;
+  }
+  weak_factory_.InvalidateWeakPtrs();
```

Comment:
please leave a comment here explaining that we want to make sure any in-progress user script that returns after the timeout is a no-op